File: Sources/MCGACore/ValueParsers.swift
import Foundation

struct UUIDParser: ContentParser {
    let name = "UUID"

    func parse(_ content: String, previousContent: String) -> [ParseResult] {
        guard content.contains("-"), let uuid = UUID(uuidString: content) else { return [] }
        let bytes = uuid.uuidBytes
        let version = Int((bytes[6] & 0xf0) >> 4)
        let versionInfo = versionDescription(version)
        let variantInfo = variantDescription(bytes[8])
        let high = bytes.prefix(8).reduce(UInt64(0)) { ($0 << 8) | UInt64($1) }
        let low = bytes.suffix(8).reduce(UInt64(0)) { ($0 << 8) | UInt64($1) }

        var details = """
        版本：\(versionInfo)
        变体：\(variantInfo)
        大写：\(uuid.uuidString.uppercased())
        URN: urn:uuid:\(uuid.uuidString.lowercased())
        高 64 位：0x\(String(format: "%016llx", high))
        低 64 位：0x\(String(format: "%016llx", low))
        """

        var timeInfo = ""
        if let date = uuidTime(version: version, bytes: bytes) {
            let formatted = ParserUtilities.utcString(from: date)
            timeInfo = "\n创建时间：\(formatted)"
            details += "\n\n时间信息:\n  创建时间：\(formatted)\n  ISO 8601: \(ParserUtilities.isoString(from: date))"
            if version == 1 {
                let mac = bytes[10...15].map { String(format: "%02x", $0) }.joined(separator: ":")
                details += "\n\n节点信息:\n  MAC 地址：\(mac)"
                if bytes[10] & 0x01 == 0x01 {
                    details += " (随机生成)"
                }
            }
        } else if [2, 3, 4, 5, 8].contains(version) {
            timeInfo = "\n时间信息：无 (该版本不包含时间戳)"
            details += "\n\n时间信息：该版本不包含时间戳"
        } else {
            timeInfo = "\n时间信息：无法解析"
        }

        return [ParseResult(parserName: name, original: content, parsed: "\(versionInfo)\(timeInfo)", details: details)]
    }

    private func versionDescription(_ version: Int) -> String {
        switch version {
        case 1: "v1 (基于时间和 MAC 地址)"
        case 2: "v2 (DCE Security)"
        case 3: "v3 (基于 MD5 哈希)"
        case 4: "v4 (随机生成)"
        case 5: "v5 (基于 SHA-1 哈希)"
        case 6: "v6 (有序时间戳)"
        case 7: "v7 (Unix 时间戳)"
        case 8: "v8 (自定义)"
        default: "未知版本"
        }
    }

    private func variantDescription(_ byte: UInt8) -> String {
        if byte & 0x80 == 0 { return "NCS 向后兼容" }
        if byte & 0xc0 == 0x80 { return "RFC 4122" }
        if byte & 0xe0 == 0xc0 { return "Microsoft 向后兼容" }
        return "保留给未来定义"
    }

    private func uuidTime(version: Int, bytes: [UInt8]) -> Date? {
        let uuidEpochDiff: UInt64 = 122_192_928_000_000_000
        let timestamp: UInt64
        switch version {
        case 1:
            let timeLow = UInt64(bytes[0]) << 24 | UInt64(bytes[1]) << 16 | UInt64(bytes[2]) << 8 | UInt64(bytes[3])
            let timeMid = UInt64(bytes[4]) << 8 | UInt64(bytes[5])
            let timeHigh = (UInt64(bytes[6]) << 8 | UInt64(bytes[7])) & 0x0fff
            timestamp = timeLow | (timeMid << 32) | (timeHigh << 48)
        case 6:
            let high = UInt64(bytes[0]) << 24 | UInt64(bytes[1]) << 16 | UInt64(bytes[2]) << 8 | UInt64(bytes[3])
            let mid = UInt64(bytes[4]) << 8 | UInt64(bytes[5])
            let low = (UInt64(bytes[6]) << 8 | UInt64(bytes[7])) & 0x0fff
            timestamp = (high << 28) | (mid << 12) | low
        case 7:
            let millis = bytes[0...5].reduce(UInt64(0)) { ($0 << 8) | UInt64($1) }
            return Date(timeIntervalSince1970: TimeInterval(millis) / 1000)
        default:
            return nil
        }
        guard timestamp >= uuidEpochDiff else { return nil }
        let unix100ns = timestamp - uuidEpochDiff
        return Date(timeIntervalSince1970: TimeInterval(unix100ns) / 10_000_000)
    }
}

struct ObjectIDParser: ContentParser {
    let name = "ObjectID"
    private let pattern = ParserUtilities.regex(#"^[0-9a-fA-F]{24}$"#)

    func parse(_ content: String, previousContent: String) -> [ParseResult] {
        guard ParserUtilities.fullMatch(pattern, content) != nil else { return [] }
        let bytes = stride(from: 0, to: content.count, by: 2).compactMap { offset -> UInt8? in
            let start = content.index(content.startIndex, offsetBy: offset)
            let end = content.index(start, offsetBy: 2)
            return UInt8(content[start..<end], radix: 16)
        }
        guard bytes.count == 12 else { return [] }
        let seconds = UInt32(bytes[0]) << 24 | UInt32(bytes[1]) << 16 | UInt32(bytes[2]) << 8 | UInt32(bytes[3])
        let date = Date(timeIntervalSince1970: TimeInterval(seconds))
        let random = ParserUtilities.hex(bytes[4..<9])
        let counter = UInt32(bytes[9]) << 16 | UInt32(bytes[10]) << 8 | UInt32(bytes[11])
        return [ParseResult(
            parserName: name,
            original: content,
            parsed: "创建时间: \(ParserUtilities.utcSecondString(from: date))\n随机值: \(random)\n计数器: \(counter)",
            details: "时间戳: \(seconds)\nISO 8601: \(ParserUtilities.isoString(from: date))"
        )]
    }
}

struct HashParser: ContentParser {
    let name = "Hash"
    private let pattern = ParserUtilities.regex(#"^[0-9a-fA-F]+$"#)

    func parse(_ content: String, previousContent: String) -> [ParseResult] {
        guard ParserUtilities.fullMatch(pattern, content) != nil else { return [] }
        let kind: String
        switch content.count {
        case 32: kind = "MD5"
        case 40: kind = "SHA-1"
        case 56: kind = "SHA-224"
        case 64: kind = "SHA-256"
        case 96: kind = "SHA-384"
        case 128: kind = "SHA-512"
        default: return []
        }
        return [ParseResult(parserName: name, original: content, parsed: "类型：\(kind)\n长度：\(content.count) hex 字符")]
    }
}

struct TimestampParser: ContentParser {
    let name = "Timestamp"

    func parse(_ content: String, previousContent: String) -> [ParseResult] {
        guard content.allSatisfy(\.isNumber), let value = Int64(content) else { return [] }
        let unit: String
        let seconds: TimeInterval
        switch content.count {
        case 10:
            unit = "秒"
            seconds = TimeInterval(value)
        case 13:
            unit = "毫秒"
            seconds = TimeInterval(value) / 1_000
        case 16:
            unit = "微秒"
            seconds = TimeInterval(value) / 1_000_000
        case 17:
            unit = "百纳秒"
            seconds = TimeInterval(value) / 10_000_000
        case 19:
            unit = "纳秒"
            seconds = TimeInterval(value) / 1_000_000_000
        default:
            return []
        }
        guard seconds >= 0, seconds <= 4_102_444_800 else { return [] }
        let date = Date(timeIntervalSince1970: seconds)
        let formatted = ParserUtilities.utcString(from: date)
        return [ParseResult(
            parserName: name,
            original: content,
            parsed: "精度：\(unit)\n时间：\(formatted)",
            details: "原始值：\(value)\n精度：\(unit)\nUTC: \(formatted)\nISO 8601: \(ParserUtilities.isoString(from: date))"
        )]
    }
}

    private let pattern = ParserUtilities.regex(#"^[A-Za-z0-9_-]{16}$"#)

    func parse(_ content: String, previousContent: String) -> [ParseResult] {
        guard ParserUtilities.fullMatch(pattern, content) != nil,
              let (data, _) = ParserUtilities.dataFromBase64Variants(content),
              data.count == 12
        else { return [] }
        let bytes = [UInt8](data)
        let pid = UInt32(bytes[0]) | UInt32(bytes[1]) << 8 | UInt32(bytes[2]) << 16 | UInt32(bytes[3]) << 24
        let unixNano = bytes[4..<12].enumerated().reduce(Int64(0)) { acc, item in
            acc | (Int64(item.element) << Int64(item.offset * 8))
        }
        let seconds = TimeInterval(unixNano) / 1_000_000_000
        let date = Date(timeIntervalSince1970: seconds)
        guard date >= Date(timeIntervalSince1970: 1_420_070_400) else { return [] }
        return [ParseResult(
            parserName: name,
            original: content,
            parsed: "时间：\(ParserUtilities.utcString(from: date))\nPID: \(pid)",
            details: "UnixNano: \(unixNano)\nISO 8601: \(ParserUtilities.isoString(from: date))"
        )]
    }
}

extension UUID {
    var uuidBytes: [UInt8] {
        withUnsafeBytes(of: uuid) { Array($0) }
    }
}

File: windows/MCGACore/ValueParsers.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.IO;
using System.Linq;
using System.Net;
using System.Text;
using System.Text.RegularExpressions;
using System.Xml.Linq;

namespace MCGA.MCGACore
{
    public class UUIDParser : IContentParser
    {
        public string Name => "UUID";
        public ParserInfo? Info => null;

        public List<ParseResult> Parse(string content, string previousContent = "")
        {
            if (!content.Contains("-") || !Guid.TryParse(content, out var guid)) return new List<ParseResult>();

            byte[] guidBytes = guid.ToByteArray();
            byte[] bytes = new byte[16];
            // Convert to Big-Endian Network Byte Order (RFC 4122)
            bytes[0] = guidBytes[3];
            bytes[1] = guidBytes[2];
            bytes[2] = guidBytes[1];
            bytes[3] = guidBytes[0];
            bytes[4] = guidBytes[5];
            bytes[5] = guidBytes[4];
            bytes[6] = guidBytes[7];
            bytes[7] = guidBytes[6];
            Array.Copy(guidBytes, 8, bytes, 8, 8);

            int version = (bytes[6] & 0xf0) >> 4;
            string versionInfo = GetVersionDescription(version);
            string variantInfo = GetVariantDescription(bytes[8]);

            ulong high = 0;
            for (int i = 0; i < 8; i++) high = (high << 8) | bytes[i];
            ulong low = 0;
            for (int i = 8; i < 16; i++) low = (low << 8) | bytes[i];

            string details = $"版本：{versionInfo}\n" +
                             $"变体：{variantInfo}\n" +
                             $"大写：{guid.ToString().ToUpperInvariant()}\n" +
                             $"URN: urn:uuid:{guid.ToString().ToLowerInvariant()}\n" +
                             $"高 64 位：0x{high:x16}\n" +
                             $"低 64 位：0x{low:x16}";

            string timeInfo = "";
            var date = GetUuidTime(version, bytes);
            if (date.HasValue)
            {
                string formatted = ParserUtilities.UtcString(date.Value);
                timeInfo = $"\n创建时间：{formatted}";
                details += $"\n\n时间信息:\n  创建时间：{formatted}\n  ISO 8601: {ParserUtilities.IsoString(date.Value)}";
                if (version == 1)
                {
                    string mac = string.Join(":", bytes.Skip(10).Take(6).Select(b => b.ToString("x2")));
                    details += $"\n\n节点信息:\n  MAC 地址：{mac}";
                    if ((bytes[10] & 0x01) == 0x01)
                    {
                        details += " (随机生成)";
                    }
                }
            }
            else if (version == 2 || version == 3 || version == 4 || version == 5 || version == 8)
            {
                timeInfo = "\n时间信息：无 (该版本不包含时间戳)";
                details += "\n\n时间信息：该版本不包含时间戳";
            }
            else
            {
                timeInfo = "\n时间信息：无法解析";
            }

            return new List<ParseResult>
            {
                new ParseResult(Name, content, $"{versionInfo}{timeInfo}", details)
            };
        }

        private string GetVersionDescription(int version)
        {
            return version switch
            {
                1 => "v1 (基于时间和 MAC 地址)",
                2 => "v2 (DCE Security)",
                3 => "v3 (基于 MD5 哈希)",
                4 => "v4 (随机生成)",
                5 => "v5 (基于 SHA-1 哈希)",
                6 => "v6 (有序时间戳)",
                7 => "v7 (Unix 时间戳)",
                8 => "v8 (自定义)",
                _ => "未知版本"
            };
        }

        private string GetVariantDescription(byte byteValue)
        {
            if ((byteValue & 0x80) == 0) return "NCS 向后兼容";
            if ((byteValue & 0xc0) == 0x80) return "RFC 4122";
            if ((byteValue & 0xe0) == 0xc0) return "Microsoft 向后兼容";
            return "保留给未来定义";
        }

        private DateTime? GetUuidTime(int version, byte[] bytes)
        {
            const ulong uuidEpochDiff = 122192928000000000UL;
            ulong timestamp = 0;
            switch (version)
            {
                case 1:
                    ulong timeLow = ((ulong)bytes[0] << 24) | ((ulong)bytes[1] << 16) | ((ulong)bytes[2] << 8) | bytes[3];
                    ulong timeMid = ((ulong)bytes[4] << 8) | bytes[5];
                    ulong timeHigh = (((ulong)bytes[6] << 8) | bytes[7]) & 0x0fffUL;
                    timestamp = timeLow | (timeMid << 32) | (timeHigh << 48);
                    break;
                case 6:
                    ulong high = ((ulong)bytes[0] << 24) | ((ulong)bytes[1] << 16) | ((ulong)bytes[2] << 8) | bytes[3];
                    ulong mid = ((ulong)bytes[4] << 8) | bytes[5];
                    ulong low = (((ulong)bytes[6] << 8) | bytes[7]) & 0x0fffUL;
                    timestamp = (high << 28) | (mid << 12) | low;
                    break;
                case 7:
                    ulong millis = 0;
                    for (int i = 0; i < 6; i++) millis = (millis << 8) | bytes[i];
                    return DateTimeOffset.FromUnixTimeMilliseconds((long)millis).UtcDateTime;
                default:
                    return null;
            }

            if (timestamp < uuidEpochDiff) return null;
            ulong unix100ns = timestamp - uuidEpochDiff;
            double seconds = (double)unix100ns / 10000000.0;
            return DateTimeOffset.FromUnixTimeSeconds(0).AddSeconds(seconds).UtcDateTime;
        }
    }

    public class ObjectIDParser : IContentParser
    {
        public string Name => "ObjectID";
        public ParserInfo? Info => null;
        private readonly Regex _pattern = ParserUtilities.GetRegex(@"^[0-9a-fA-F]{24}$");

        public List<ParseResult> Parse(string content, string previousContent = "")
        {
            if (ParserUtilities.FullMatch(_pattern, content) == null) return new List<ParseResult>();

            byte[] bytes = new byte[12];
            for (int i = 0; i < 12; i++)
            {
                bytes[i] = byte.Parse(content.Substring(i * 2, 2), NumberStyles.HexNumber);
            }

            uint seconds = ((uint)bytes[0] << 24) | ((uint)bytes[1] << 16) | ((uint)bytes[2] << 8) | bytes[3];
            var date = DateTimeOffset.FromUnixTimeSeconds(seconds).UtcDateTime;
            string random = ParserUtilities.Hex(bytes.Skip(4).Take(5).ToArray());
            uint counter = ((uint)bytes[9] << 16) | ((uint)bytes[10] << 8) | bytes[11];

            return new List<ParseResult>
            {
                new ParseResult(
                    Name,
                    content,
                    $"创建时间: {ParserUtilities.UtcSecondString(date)}\n随机值: {random}\n计数器: {counter}",
                    $"时间戳: {seconds}\nISO 8601: {ParserUtilities.IsoString(date)}"
                )
            };
        }
    }

    public class HashParser : IContentParser
    {
        public string Name => "Hash";
        public ParserInfo? Info => null;
        private readonly Regex _pattern = ParserUtilities.GetRegex(@"^[0-9a-fA-F]+$");

        public List<ParseResult> Parse(string content, string previousContent = "")
        {
            if (ParserUtilities.FullMatch(_pattern, content) == null) return new List<ParseResult>();

            string kind = content.Length switch
            {
                32 => "MD5",
                40 => "SHA-1",
                56 => "SHA-224",
                64 => "SHA-256",
                96 => "SHA-384",
                128 => "SHA-512",
                _ => ""
            };

            if (string.IsNullOrEmpty(kind)) return new List<ParseResult>();

            return new List<ParseResult>
            {
                new ParseResult(Name, content, $"类型：{kind}\n长度：{content.Length} hex 字符")
            };
        }
    }

    public class TimestampParser : IContentParser
    {
        public string Name => "Timestamp";
        public ParserInfo? Info => null;

        public List<ParseResult> Parse(string content, string previousContent = "")
        {
            if (!content.All(char.IsDigit) || !long.TryParse(content, out long value)) return new List<ParseResult>();

            string unit;
            double seconds;
            switch (content.Length)
            {
                case 10:
                    unit = "秒";
                    seconds = value;
                    break;
                case 13:
                    unit = "毫秒";
                    seconds = value / 1000.0;
                    break;
                case 16:
                    unit = "微秒";
                    seconds = value / 1000000.0;
                    break;
                case 17:
                    unit = "百纳秒";
                    seconds = value / 10000000.0;
                    break;
                case 19:
                    unit = "纳秒";
                    seconds = value / 1000000000.0;
                    break;
                default:
                    return new List<ParseResult>();
            }

            if (seconds < 0 || seconds > 4102444800.0) return new List<ParseResult>();

            var date = DateTimeOffset.FromUnixTimeSeconds(0).AddSeconds(seconds).UtcDateTime;
            string formatted = ParserUtilities.UtcString(date);

            return new List<ParseResult>
            {
                new ParseResult(
                    Name,
                    content,
                    $"精度：{unit}\n时间：{formatted}",
                    $"原始值：{value}\n精度：{unit}\nUTC: {formatted}\nISO 8601: {ParserUtilities.IsoString(date)}"
                )
            };
        }
    }

    public class HTTPStatusParser : IContentParser
    {
        public string Name => "HTTP Status";
        public ParserInfo? Info => null;

        public List<ParseResult> Parse(string content, string previousContent = "")
        {
            if (content.Length != 3 || !int.TryParse(content, out int code) || !_phrases.TryGetValue(code, out string? phrase))
            {
                return new List<ParseResult>();
            }

            string klass = code switch
            {
                >= 100 and < 200 => "信息响应",
                >= 200 and < 300 => "成功",
                >= 300 and < 400 => "重定向",
                >= 400 and < 500 => "客户端错误",
                >= 500 and < 600 => "服务端错误",
                _ => "未知"
            };

            return new List<ParseResult>
            {
                new ParseResult(Name, content, $"{code} {phrase}\n类型：{klass}")
            };
        }

        private readonly Dictionary<int, string> _phrases = new Dictionary<int, string>
        {
            { 100, "Continue" }, { 101, "Switching Protocols" }, { 102, "Processing" }, { 103, "Early Hints" },
            { 200, "OK" }, { 201, "Created" }, { 202, "Accepted" }, { 203, "Non-Authoritative Information" }, { 204, "No Content" }, { 205, "Reset Content" }, { 206, "Partial Content" }, { 207, "Multi-Status" }, { 208, "Already Reported" }, { 226, "IM Used" },
            { 300, "Multiple Choices" }, { 301, "Moved Permanently" }, { 302, "Found" }, { 303, "See Other" }, { 304, "Not Modified" }, { 305, "Use Proxy" }, { 307, "Temporary Redirect" }, { 308, "Permanent Redirect" },
            { 400, "Bad Request" }, { 401, "Unauthorized" }, { 402, "Payment Required" }, { 403, "Forbidden" }, { 404, "Not Found" }, { 405, "Method Not Allowed" }, { 406, "Not Acceptable" }, { 407, "Proxy Authentication Required" }, { 408, "Request Timeout" }, { 409, "Conflict" }, { 410, "Gone" }, { 411, "Length Required" }, { 412, "Precondition Failed" }, { 413, "Content Too Large" }, { 414, "URI Too Long" }, { 415, "Unsupported Media Type" }, { 416, "Range Not Satisfiable" }, { 417, "Expectation Failed" }, { 418, "I'm a teapot" }, { 421, "Misdirected Request" }, { 422, "Unprocessable Content" }, { 423, "Locked" }, { 424, "Failed Dependency" }, { 425, "Too Early" }, { 426, "Upgrade Required" }, { 428, "Precondition Required" }, { 429, "Too Many Requests" }, { 431, "Request Header Fields Too Large" }, { 451, "Unavailable For Legal Reasons" },
            { 500, "Internal Server Error" }, { 501, "Not Implemented" }, { 502, "Bad Gateway" }, { 503, "Service Unavailable" }, { 504, "Gateway Timeout" }, { 505, "HTTP Version Not Supported" }, { 506, "Variant Also Negotiates" }, { 507, "Insufficient Storage" }, { 508, "Loop Detected" }, { 510, "Not Extended" }, { 511, "Network Authentication Required" }
        };
    }

    public class NumberBaseParser : IContentParser
    {
        public string Name => "Number Base";
        public ParserInfo? Info => null;

        public List<ParseResult> Parse(string content, string previousContent = "")
        {
            string lower = content.Trim().ToLowerInvariant();
            ulong? parsed = null;
            int sourceBase = 10;

            try
            {
                if (lower.StartsWith("0x"))
                {
                    parsed = Convert.ToUInt64(lower.Substring(2), 16);
                    sourceBase = 16;
                }
                else if (lower.StartsWith("0b"))
                {
                    parsed = Convert.ToUInt64(lower.Substring(2), 2);
                    sourceBase = 2;
                }
                else if (lower.StartsWith("0o"))
                {
                    parsed = Convert.ToUInt64(lower.Substring(2), 8);
                    sourceBase = 8;
                }
                else if (content.Length <= 9 && content.All(char.IsDigit))
                {
                    parsed = Convert.ToUInt64(content, 10);
                    sourceBase = 10;
                }
            }
            catch
            {
                return new List<ParseResult>();
            }

            if (!parsed.HasValue) return new List<ParseResult>();
            ulong value = parsed.Value;

            var lines = new List<string>
            {
                $"输入进制：{sourceBase}",
                $"DEC: {value}",
                $"HEX: 0x{Convert.ToString((long)value, 16).ToUpperInvariant()}",
                $"OCT: 0o{Convert.ToString((long)value, 8)}",
                $"BIN: 0b{Convert.ToString((long)value, 2)}"
            };

            return new List<ParseResult>
            {
                new ParseResult(Name, content, string.Join("\n", lines))
            };
        }
    }

    public class URLParser : IContentParser
    {
        public string Name => "URL";
        public ParserInfo? Info => null;

        public List<ParseResult> Parse(string content, string previousContent = "")
        {
            if (content.Length > 8192 || !content.Contains("://")) return new List<ParseResult>();

            try
            {
                var uri = new Uri(content);
                var lines = new List<string>
                {
                    $"Scheme: {uri.Scheme}",
                    $"Host: {uri.Host}"
                };

                if (!uri.IsDefaultPort)
                {
                    lines.Add($"Port: {uri.Port}");
                }
                if (!string.IsNullOrEmpty(uri.AbsolutePath))
                {
                    lines.Add($"Path: {uri.AbsolutePath}");
                }
                if (!string.IsNullOrEmpty(uri.Fragment))
                {
                    lines.Add($"Fragment: {uri.Fragment.TrimStart('#')}");
                }

                var query = uri.Query;
                if (!string.IsNullOrEmpty(query))
                {
                    var pairs = query.TrimStart('?').Split('&');
                    if (pairs.Length > 0 && pairs[0].Length > 0)
                    {
                        lines.Add("");
                        lines.Add("Query:");
                        foreach (var pair in pairs)
                        {
                            int eq = pair.IndexOf('=');
                            if (eq >= 0)
                            {
                                string k = pair.Substring(0, eq);
                                string v = pair.Substring(eq + 1);
                                lines.Add($"  {k} = {Uri.UnescapeDataString(v)}");
                            }
                            else
                            {
                                lines.Add($"  {pair} = ");
                            }
                        }
                    }
                }

                string decoded = Uri.UnescapeDataString(content);
                string details = decoded == content ? string.Join("\n", lines) : $"{string.Join("\n", lines)}\n\nDecoded:\n{decoded}";

                return new List<ParseResult>
                {
                    new ParseResult(Name, content, string.Join("\n", lines), details)
                };
            }
            catch
            {
                return new List<ParseResult>();
            }
        }
    }

    public class HTMLEntityParser : IContentParser
    {
        public string Name => "HTML Entity";
        public ParserInfo? Info => null;
        private readonly Regex _pattern = ParserUtilities.GetRegex(@"&(?:#[0-9]+|#x[0-9a-fA-F]+|[a-zA-Z0-9]+);");

        public List<ParseResult> Parse(string content, string previousContent = "")
        {
            if (ParserUtilities.FullMatch(_pattern, content) == null && !_pattern.IsMatch(content)) return new List<ParseResult>();

            string decoded = WebUtility.HtmlDecode(content);
            if (decoded == content) return new List<ParseResult>();

            return new List<ParseResult>
            {
                new ParseResult(Name, content, decoded, $"原始长度：{content.Length}\n解码长度：{decoded.Length}")
            };
        }
    }

    public class TOMLParser : IContentParser
    {
        public string Name => "TOML";
        public ParserInfo? Info => null;
        private readonly Regex _assignment = ParserUtilities.GetRegex(@"^[A-Za-z0-9_.-]+\s*=\s*.+");

        public List<ParseResult> Parse(string content, string previousContent = "")
        {
            string trimmed = content.Trim();
            if (Encoding.UTF8.GetByteCount(trimmed) > 64 * 1024) return new List<ParseResult>();
            if (trimmed.StartsWith("{") || trimmed.StartsWith("[")) return new List<ParseResult>();
            if (!LooksLikeTOML(trimmed)) return new List<ParseResult>();

            var lines = trimmed.Split(new[] { '\r', '\n' }, StringSplitOptions.None)
                .Select(FormatLine)
                .ToList();

            string formatted = string.Join("\n", lines).Trim();
            if (string.IsNullOrEmpty(formatted)) return new List<ParseResult>();

            return new List<ParseResult>
            {
                new ParseResult(Name, content, $"TOML  大小：{Encoding.UTF8.GetByteCount(trimmed)} 字节", formatted)
            };
        }

        private bool LooksLikeTOML(string content)
        {
            var lines = content.Split('\n');
            foreach (var line in lines)
            {
                string tr = line.Trim();
                if (tr.StartsWith("#")) continue;

                int hash = tr.IndexOf('#');
                string uncommented = hash >= 0 ? tr.Substring(0, hash) : tr;
                uncommented = uncommented.Trim();

                if (tr.StartsWith("[") && tr.EndsWith("]")) return true;
                if (ParserUtilities.FullMatch(_assignment, uncommented) != null) return true;
            }
            return false;
        }

        private string FormatLine(string line)
        {
            string tr = line.Trim();
            if (string.IsNullOrEmpty(tr) || tr.StartsWith("#")) return line;

            int equals = tr.IndexOf('=');
            if (equals < 0) return line;

            string key = tr.Substring(0, equals).Trim();
            string val = tr.Substring(equals + 1).Trim();
            return $"{key} = {val}";
        }
    }

    public class XMLFormatParser : IContentParser
    {
        public string Name => "XML";
        public ParserInfo? Info => null;

        public List<ParseResult> Parse(string content, string previousContent = "")
        {
            string trimmed = content.Trim();
            if (Encoding.UTF8.GetByteCount(trimmed) > 128 * 1024) return new List<ParseResult>();
            if (!trimmed.StartsWith("<") || !trimmed.EndsWith(">")) return new List<ParseResult>();

            try
            {
                var doc = XDocument.Parse(trimmed);
                string formatted = doc.ToString();
                string root = doc.Root?.Name.LocalName ?? "unknown";
                int bytes = Encoding.UTF8.GetByteCount(trimmed);

                return new List<ParseResult>
                {
                    new ParseResult(Name, content, $"Root: {root}\n大小：{bytes} 字节", formatted)
                };
            }
            catch
            {
                return new List<ParseResult>();
            }
        }
    }
}
